File: object/Astar.py
import sys
from Rules import *
from State import *

class 	Astar:

# Creates an object to find the terminal state by the specified rules.
	def 	__init__( self, rules ):
		self.closedStates = 0
		self.rules = rules

# Applies the algorithm A * to find the shortest path to the terminal state from the indicated.
# return @param a sequence of states from a given state to a terminal state.
	def 	search( self, startState ):
		closeList = list()
		openList = list()
		openList.append( startState )
		startState.setG( 0 )
		startState.setH( self.rules.getH( startState ))
		sizeLine = startState.getMatrixObject().getSize()
		size = sizeLine * sizeLine
		while (len(openList)!= 0):
			current,i = self.getStateWithMinF(openList)
			if 	self.rules.isTerminate(current):
				listState = list()
				return self.completeSolution(current,listState)
			openList.pop(i)
			closeList.append(current) 
			neighborsListState = self.rules.getNeighbors(current)
			for next in neighborsListState:
				if ((self.find(next, closeList, size, sizeLine)) == True):
					continue
				gScope = current.getG() + 1
				isGBetter = False
				if ((self.find(next, openList, size, sizeLine) == False)):
					next.setH(self.rules.getH(next))
					openList.append(next)
					isGBetter = True		
				else:
					isGBetter = gScope <= next.getG()
				if (isGBetter == True):
					next.setStateParent(current)
					next.setG(gScope)
		return 0;
 


	def 	find(self, state, listState, size, sizeLine):
		s = state.getMatrixArray()
		for currentState in listState:
			res = 0
			tmp = currentState.getMatrixArray()
			i = 0
			while (i < sizeLine):
				j = 0
				flagBreak = 0
				while j < sizeLine:
					if (s[i][j] == tmp[i][j]):
						res += 1
					else:
						flagBreak = 1
						break
					j += 1
				i += 1
				if (flagBreak == 1):
					break 
			if (res == size):
				return True
		return (False)

	def 	getStateWithMinF( self, openList ):
		min = sys.maxsize - 1
		i = 0
		i_res = 0
		for state in openList:
			if state.getF() < min:
				min = state.getF()
				i_res = i
			i += 1
		res = copy.deepcopy(openList[i_res])
		return res, i_res

	def 	completeSolution( self, terminate, listState ):
		listState.append(terminate.getMatrixArray())
		if (terminate.countParent != 0):
			return self.completeSolution(terminate.getStateParent(), listState)
		else:
			return listState

File: object/astar.java
package ru.dokwork.algorithms.astar;

import java.util.*;

/**
 * Реализует алгоритм поиска решения А*.
 */
public class Astar<TState extends State, TRules extends Rules<TState>> {

    /**
     * Применяет алгоритм А* для поиска крадчайшего пути до терминального
     * состоянияот указанного.
     *
     * @param startState - начальное состояние.
     * @return последовательность состояний от заданного до терминального.
     */
    public Collection<State> search(TState startState) {
        LinkedList<Integer> close = new LinkedList<Integer>();
        LinkedList<TState> open = new LinkedList<TState>();
        open.add(startState);
        startState.setG(0);
        startState.setH(rules.getH(startState));

        while (!open.isEmpty()) {

            TState x = getStateWithMinF(open);
           
            if (rules.isTerminate(x)) {
                closedStates = close.size();
                return completeSolution(x);
            }

            open.remove(x);
            close.add(x.hashCode());

            List<TState> neighbors = rules.getNeighbors(x);
            
            for (TState neighbor : neighbors) {
                if (close.contains(neighbor.hashCode())) {
                    continue;
                }
                int g = x.getG() + rules.getDistance(x, neighbor);
                boolean isGBetter;

                if (!open.contains(neighbor)) {
                    neighbor.setH(rules.getH(neighbor));
                    open.add(neighbor);
                    isGBetter = true;
                } else {
                    isGBetter = g < neighbor.getG();
                }
                if (isGBetter) {
                    neighbor.setParent(x);
                    neighbor.setG(g);
                }
            }

        }
        return null;
    }

    public int getClosedStatesCount() {
        return closedStates;
    }

    /**
     * Создает объект для поиска терминального состояния по указанным правилам.
     *
     * @param rules правила, в соответствии с которыми будет производиться поиск
     *              терминального состояния.
     */
    public Astar(TRules rules) {
        if (rules == null) {
            throw new IllegalArgumentException("Rules can`t be null.");
        }
        this.rules = rules;
    }

    /**
     * Находит вершину в списке open с наименьшим значением веса.
     *
     * @param open список открытых вершин.
     * @return вершину с наименьшим весом.
     */
    private TState getStateWithMinF(Collection<TState> open) {
        TState res = null;
        int min = Integer.MAX_VALUE;
        for (TState state : open) {
            if (state.getF() < min) {
                min = state.getF();
                res = state;
            }
        }
        return res;
    }

    /**
     * Составляет последовательность состояний пройденных от начального
     * состояния до конечного.
     *
     * @param terminate найденное конечное состояние.
     * @return последовательность состояний пройденных от начального
     * состояния до конечного.
     */
    private Collection<State> completeSolution(TState terminate) {
        LinkedList<State> path = new LinkedList<State>();
        State c = terminate;
        while (c != null) {
            path.addFirst(c);
            c = c.getParent();
        }
        return path;
    }

    private TRules rules;
    private int closedStates = 0;
}


















public static List<Point> FindPath(int[,] field, Point start, Point goal)
{
  // Шаг 1.
  var closedSet = new Collection<PathNode>();
  var openSet = new Collection<PathNode>();
  // Шаг 2.
  PathNode startNode = new PathNode()
  {
    Position = start,
    CameFrom = null,
    PathLengthFromStart = 0,
    HeuristicEstimatePathLength = GetHeuristicPathLength(start, goal)
  };

  openSet.Add(startNode);
  while (openSet.Count > 0)
  {
    // Шаг 3.
    var currentNode = openSet.OrderBy(node => 
      node.EstimateFullPathLength).First();
    // Шаг 4.
    if (currentNode.Position == goal)
      return GetPathForNode(currentNode);


    // Шаг 5.
    openSet.Remove(currentNode);
    closedSet.Add(currentNode);


    // Шаг 6.
    foreach (var neighbourNode in GetNeighbours(currentNode, goal, field))
    {
      // Шаг 7.
      if (closedSet.Count(node => node.Position == neighbourNode.Position) > 0)
        continue;
     

     var openNode = openSet.FirstOrDefault(node => node.Position == neighbourNode.Position);
      // Шаг 8.
      if (openNode == null)
        openSet.Add(neighbourNode);
      else
        if (openNode.PathLengthFromStart > neighbourNode.PathLengthFromStart)
        {
          // Шаг 9.
          openNode.CameFrom = currentNode;
          openNode.PathLengthFromStart = neighbourNode.PathLengthFromStart;
        }
    }
  }
  // Шаг 10.
  return null;
}
